optimize the decomposition of aten.native_group_norm

jiayisunx opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #144733

Summary:
Optimize the decomposition of aten.native_group_norm. Reduce unnecessary repeated operations by changing the order of operations for `mean`, `rstd`, `weight`, `bias `and `input`, which can improve performance when `flattened_inner_size `is large.

The original decomposition:
1. compute `mean `and `rstd`,
2. out = (x - mean) * rstd, compute in the range [N, C, *],
3. out = out * weight + bias, compute in the range [N, C, *],

The new decomposition:
1. compute `mean `and `rstd`,
2. new_weight = rstd * weight, new_bias = - mean * rstd * weight + bias, compute in the range [N, C],
3. out = out * new_weight + new_bias, compute in the range [N, C, *],

I tested the Inductor performance benchmark with this PR on both CPU and A100. On CPU, two torchbench models(functorch_dp_cifar10 and opacus_cifar10) have about 25% performance improvement, and two diffusion models(Stable Diffusion and Latent Consistency Model(LCM)) have about 2% performance improvement. On A100, no performance gains or regressions were seen.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

jiayisunx committed:
Update

[ghstack-poisoned]

jiayisunx committed:
Update

[ghstack-poisoned]

jiayisunx committed:
Update

[ghstack-poisoned]

jiayisunx committed:
Update

[ghstack-poisoned]

jiayisunx committed:
Update

[ghstack-poisoned]

jiayisunx committed:
Update

[ghstack-poisoned]

jiayisunx committed:
Update

[ghstack-poisoned]

jiayisunx committed:
Update

[ghstack-poisoned]

jiayisunx committed:
Update

[ghstack-poisoned]

jiayisunx committed:
Update

[ghstack-poisoned]

jiayisunx committed:
Update

[ghstack-poisoned]

leslie-fang-intel reviewed (commented):
Please update the performance impact of this optimization for both CPP and Triton Backend.

jiayisunx commented:
> Please update the performance impact of this optimization for both CPP and Triton Backend.

Added, thanks for your comments!

jansel commented on test/inductor/test_cpu_repro.py:
```diff
@@ -4053,7 +4053,7 @@ def forward(self, x):
                 actual, code = run_and_get_cpp_code(compiled_m, x)
                 self.assertEqual(expected, actual)
                 # 2 generated kernels (one for var_mean, the other for result)
```
Comment is out of date

jansel commented on torch/_refs/__init__.py:
```diff
@@ -3192,27 +3192,44 @@ def native_group_norm(
-    if unsqueeze_bias is not None:
-        out = out + unsqueeze_bias
+    if weight is not None:
+        w = w.as_strided([batch_size, num_channels], [num_channels, 1])
+        b = b.as_strided([batch_size, num_channels], [num_channels, 1])
```
Here you are assuming w/b are contiguous.  Won't this be incorrect if I pass in tensors with non-standard strides?

jiayisunx replied:
You are right. I modified this line of code to get the current correct w/b stride from the original w/b stride.

jiayisunx committed:
Update

[ghstack-poisoned]

jiayisunx committed:
Update

[ghstack-poisoned]

jiayisunx committed:
Update

[ghstack-poisoned]

jansel commented on torch/_refs/__init__.py:
```diff
@@ -3192,27 +3192,44 @@ def native_group_norm(
+        b = b + bias_reshaped
+
+    if input.device.type == "cpu" and weight is not None:
+        w = w.as_strided([batch_size, num_channels], [w.stride()[0], w.stride()[2]])
+        b = b.as_strided([batch_size, num_channels], [b.stride()[0], b.stride()[2]])
```
Again, this code will be incorrect for some strides.  Here you are assuming stride[0] and stride[1] are contiguous.

```py
>>> contig = torch.arange(2*2*2).reshape(2,2,2)
>>> transposed = torch.empty_strided((2,2,2), (2,1,4), dtype=torch.int64).copy_(contig)
>>> contig
tensor([[[0, 1],
         [2, 3]],

        [[4, 5],
         [6, 7]]])
>>> transposed
tensor([[[0, 1],
         [2, 3]],

        [[4, 5],
         [6, 7]]])
>>> contig.as_strided([2, 4], [contig.stride(0), contig.stride(2)])
tensor([[0, 1, 2, 3],
        [4, 5, 6, 7]])
>>> transposed.as_strided([2, 4], [transposed.stride(0), transposed.stride(2)])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: setStorage: sizes [2, 4], strides [2, 4], storage offset 0, and itemsize 8 requiring a storage size of 120 are out of bounds for storage of size 64
```

You could add a call to `.contiguous()` for force a copy if the strides are non-standard.

jansel replied:
you could get the same problem by slicing a contiguous input

jiayisunx replied:
Done, thanks for your comments!

jiayisunx committed:
Update

[ghstack-poisoned]

jiayisunx commented:
@pytorchbot merge

desertfire commented:
@pytorchbot revert -m "Cause TIMM pass rate regression on H100, see https://hud.pytorch.org/benchmark/compilers?dashboard=torchinductor&startTime=Thu%2C%2020%20Feb%202025%2020%3A53%3A55%20GMT&stopTime=Thu%2C%2027%20Feb%202025%2020%3A53%3A55%20GMT&granularity=hour&mode=training&dtype=amp&deviceName=cuda%20(h100)&lBranch=main&lCommit=4216478250e08e950fdd090fc23a1b270c520cc4&rBranch=main&rCommit=4986f0f52eb871cdb91b8124ee162cfe622b8688" -c nosignal

jansel reviewed (changes requested):
Re-request review once you have fixed the regression and done a new benchmark run.

jiayisunx committed:
Update

[ghstack-poisoned]

jiayisunx committed:
Update

[ghstack-poisoned]

jiayisunx committed:
Update

[ghstack-poisoned]

jiayisunx committed:
Update

[ghstack-poisoned]

jiayisunx commented:
Hi @jansel , I have fixed the TIMM pass rate regression issue mentioned above, see https://hud.pytorch.org/benchmark/compilers?dashboard=torchinductor&startTime=Mon%2C%2010%20Mar%202025%2001%3A49%3A48%20GMT&stopTime=Mon%2C%2017%20Mar%202025%2001%3A49%3A48%20GMT&granularity=hour&mode=training&dtype=amp&deviceName=cuda%20(h100)&lBranch=gh/jiayisunx/51/orig&lCommit=a99a52bbd7fe2505936ccc0552da557c8376889f&rBranch=main&rCommit=924a247fbbe32063966533535b1cc0cf474ef1ed. I modified this PR, currently this PR only affects CPU. Could you please review it again? Thanks!

jiayisunx committed:
Update

[ghstack-poisoned]

jiayisunx commented:
@pytorchbot merge
